kube/src/api/dynamic.rs

```diff
@@ -1,260 +1,143 @@
-    pub fn try_into_api<K: Meta>(self, client: Client) -> Result<Api<K>> {
-        let resource = Resource::try_from(self)?;
-        Ok(Api {
-            client,
-            resource,
-            phantom: iter::empty(),
-        })
-    }
-}
-
-impl TryFrom<DynamicResource> for Resource {
-    type Error = crate::Error;
-
-    fn try_from(rb: DynamicResource) -> Result<Self> {
-        if rb.version.is_none() {
-            return Err(Error::DynamicResource(format!(
-                "DynamicResource '{}' must have a version",
-                rb.kind
+    /// Set the api group, version, and kind for a resource
+    pub fn gvk(group_: &str, version_: &str, kind_: &str) -> Result<Self> {
+        let version = version_.to_string();
+        let group = group_.to_string();
+        let kind = kind_.to_string();
+        let api_version = if group.is_empty() {
+            version.to_string()
+        } else {
+            format!("{}/{}", group, version)
+        };
+        if version.is_empty() {
+            return Err(Error::DynamicType(format!(
+                "GroupVersionKind '{}' must have a version",
+                kind
             )));
         }
-        if rb.group.is_none() {
-            return Err(Error::DynamicResource(format!(
-                "DynamicResource '{}' must have a group (can be empty string)",
-                rb.kind
+        if kind.is_empty() {
+            return Err(Error::DynamicType(format!(
+                "GroupVersionKind '{}' must have a kind",
+                kind
             )));
         }
-        let version = rb.version.unwrap();
-        let group = rb.group.unwrap();
-
         Ok(Self {
-            api_version: if group.is_empty() {
-                version.clone()
-            } else {
-                format!("{}/{}", group, version)
-            },
-            kind: rb.kind,
-            version,
             group,
-            namespace: rb.namespace,
+            version,
+            kind,
+            api_version,
         })
     }
 }
 
-/// Cheaply `Clone`-able string.
-// `String`: cloning requires additional allocation.
-// `StringRef::Dynamic`: cloning is one atomic operation.
-// `StringRef::Static`: cloning is essentially free.
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-enum StringRef {
-    /// Should be used when string is known during compilation.
-    Static(&'static str),
-    /// Should be used when string is only known at runtime
-    Dynamic(Arc<str>),
-}
-
-impl StringRef {
-    fn as_str(&self) -> &str {
-        match self {
-            StringRef::Static(s) => *s,
-            StringRef::Dynamic(s) => &*s,
-        }
-    }
-}
+/// A dynamic representation of a kubernetes resource
+///
+/// This will work with any non-list type object.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug)]
+pub struct DynamicObject {
+    /// The type fields, not always present
+    #[serde(flatten, default)]
+    pub types: Option<TypeMeta>,
```

Comment:
added an optional typemeta so that it can be used in both ser/de operations.
unfortunately TypeMeta have to be optional here because when we get listable results from the `dynamic_api`, we get no TypeMeta within the .items.